test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/TransportLayerTests/RequestMessageArgsTests.cs

```diff
@@ -0,0 +1,121 @@
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+
+            _context = new Container(_baseUri)
+            {
+                HttpClientFactory = HttpClientFactory
+            };
+
+            ResetDataSource();
+        }
+
+        [Fact]
+        public async Task AddingCustomHeaderToRequest_ShouldSucceed()
+        {
+            string headerName = "MyNewHeader";
+            string headerValue = "MyNewHeaderValue";
+
+            Func<DataServiceClientRequestMessageArgs, DataServiceClientRequestMessage> configureRequest =
+                args =>
+                {
+                    args.Headers.Add(headerName, headerValue);
+                    return new HttpClientRequestMessage(args);
+                };
+
+            this.UpdateContext(configureRequest);
+
+            DataServiceQuery<ClientEndToEndModel.Product> query = _context.Products;
+            await query.ExecuteAsync();
+
+            Assert.NotNull(this.lastRequestMessage);
+            var header = this.lastRequestMessage.Headers.SingleOrDefault(h => h.Key == headerName);
+            Assert.Equal(headerValue, header.Value);
+        }
+
+        [Fact]
+        public async Task ChangingMergeToPatch_ShouldUpdateRequestMethod()
+        {
+            Func<DataServiceClientRequestMessageArgs, DataServiceClientRequestMessage> configureRequest =
+                args =>
+                {
+                    if (args.Method == "MERGE")
+                    {
+                        var newArgs = new DataServiceClientRequestMessageArgs("PATCH", args.RequestUri, true, args.Headers);
+
+                        // Ensure compatibility with V3
+                        newArgs.Headers.Remove("DataServiceVersion");
+                        newArgs.Headers.Add("DataServiceVersion", "3.0");
+
+                        return new HttpClientRequestMessage(newArgs);
+                    }
+
+                    return new HttpClientRequestMessage(args);
+                };
+
+            this.UpdateContext(configureRequest);
+
+            var query = _context.Products.Take(1) as DataServiceQuery<ClientEndToEndModel.Product>;
+            var product = (await query.ExecuteAsync()).Single();
+
+            this.lastRequestMessage = null;
+            product.Description = "New Description " + Guid.NewGuid();
+            _context.UpdateObject(product);
+            await _context.SaveChangesAsync();
+
+            Assert.NotNull(this.lastRequestMessage);
+            Assert.Equal("PATCH", this.lastRequestMessage.Method);
+        }
+
+        private Container UpdateContext(Func<DataServiceClientRequestMessageArgs, DataServiceClientRequestMessage> onMessageCreate)
+        {
+            _context.SendingRequest2 += (obj, args) => this.lastRequestMessage = args.RequestMessage;
+            _context.Configurations.RequestPipeline.OnMessageCreating = onMessageCreate;

```

Comment:
Let's include this in each test so that it's clear what's happening. It obscures a lot of important detail from the test. It also doesn't seem necessary to set `lastRequestMessage` as field of the class if its lifetime is the test method.